# Loan Default Prediction - Classification

## Part 0b. Data Aggregation

**This notebook contains code to aggregate separate data files into one master data file, after checking and ensuring the columns and rows from each file is consistent with each other.**

---


<a id = 'toc'></a>
**Table of Contents**
1. [Check sample data files - individual file format & column alignment](#sample)
    1. [data file 3a: 2007-2011](#2007)
    2. [data file 3b: 2012-2013](#2012)
    3. [data file 2016Q1](#2016q1)
    4. [data file 2019Q3](#2019q3)
    
2. [Aggregate all data files into one](#agg)

3. [Notes](#notes)

---

In [2]:
import numpy as np
import pandas as pd
pd.set_option('max_columns', 100)
pd.set_option('max_rows', 160)

import os

In [3]:
files = os.listdir("../data/raw")

In [4]:
len(files)

20

In [28]:
files = sorted(files)
for i in range(len(files)):
    print (i, files[i])

0 .DS_Store
1 LoanStats3a_securev1.csv.zip
2 LoanStats3b_securev1.csv.zip
3 LoanStats3c_securev1.csv.zip
4 LoanStats3d_securev1.csv.zip
5 LoanStats_securev1_2016Q1.csv.zip
6 LoanStats_securev1_2016Q2.csv.zip
7 LoanStats_securev1_2016Q3.csv.zip
8 LoanStats_securev1_2016Q4.csv.zip
9 LoanStats_securev1_2017Q1.csv.zip
10 LoanStats_securev1_2017Q2.csv.zip
11 LoanStats_securev1_2017Q3.csv.zip
12 LoanStats_securev1_2017Q4.csv.zip
13 LoanStats_securev1_2018Q1.csv.zip
14 LoanStats_securev1_2018Q2.csv.zip
15 LoanStats_securev1_2018Q3.csv.zip
16 LoanStats_securev1_2018Q4.csv.zip
17 LoanStats_securev1_2019Q1.csv.zip
18 LoanStats_securev1_2019Q2.csv.zip
19 LoanStats_securev1_2019Q3.csv.zip


[**back to top**](#toc)

<a id = 'sample'></a>
### 1. Check sample data files for single file format and column consistency

<a id = '2007'></a>
#### 1.A. Check data file for 2007-2011

In [8]:
%%time
df_2007_2011 = pd.read_csv(dpath+'{}'.format(files[1]), low_memory = False)

CPU times: user 3.18 s, sys: 382 ms, total: 3.56 s
Wall time: 3.6 s


In [9]:
df_2007_2011.head()

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
id      member_id loan_amnt funded_amnt funded_amnt_inv term       int_rate installment grade sub_grade emp_title           emp_length home_ownership annual_inc verification_status issue_d  loan_status pymnt_plan url                                                desc                                               purpose        title                zip_code addr_state dti   delinq_2yrs earliest_cr_line fico_range_low fico_range_high inq_last_6mths mths_since_last_delinq mths_since_last_record open_acc pub_rec revol_bal revol_util total_acc initial_list_status out_prncp out_prncp_inv total_pymnt        total_pymnt_inv total_rec_prncp total_rec_int total_rec_late_fee recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d last_fico_range_high last_fico_range_low collections_12_mths_ex_med mths_since_last_major_derog policy_code application_type annual_inc_joint dti_joint verification_status_joint acc_now_delinq tot_coll_amt tot_cur_bal open_acc_6m open_act_il open_il_12m open_il_24m mths_since_rcnt_il total_bal_il il_util open_rv_12m open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m acc_open_past_24mths avg_cur_bal bc_open_to_buy bc_util chargeoff_within_12_mths delinq_amnt mo_sin_old_il_acct mo_sin_old_rev_tl_op mo_sin_rcnt_rev_tl_op mo_sin_rcnt_tl mort_acc mths_since_recent_bc mths_since_recent_bc_dlq mths_since_recent_inq mths_since_recent_revol_delinq num_accts_ever_120_pd num_actv_bc_tl num_actv_rev_tl num_bc_sats num_bc_tl num_il_tl num_op_rev_tl num_rev_accts num_rev_tl_bal_gt_0 num_sats num_tl_120dpd_2m num_tl_30dpd num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit total_il_high_credit_limit revol_bal_joint sec_app_fico_range_low sec_app_fico_range_high sec_app_earliest_cr_line sec_app_inq_last_6mths sec_app_mort_acc sec_app_open_acc sec_app_revol_util sec_app_open_act_il sec_app_num_rev_accts sec_app_chargeoff_within_12_mths sec_app_collections_12_mths_ex_med sec_app_mths_since_last_major_derog hardship_flag hardship_type hardship_reason hardship_status deferral_term hardship_amount hardship_start_date hardship_end_date payment_plan_start_date hardship_length hardship_dpd hardship_loan_status orig_projected_additional_accrued_interest hardship_payoff_balance_amount hardship_last_payment_amount debt_settlement_flag debt_settlement_flag_date settlement_status settlement_date settlement_amount settlement_percentage                                    settlement_term                              
1077501 NaN       5000      5000        4975             36 months  10.65%  162.87      B     B2        NaN                 10+ years  RENT           24000      Verified            Dec-2011 Fully Paid  n          https://lendingclub.com/browse/loanDetail.actio...   Borrower added on 12/22/11 > I need to upgrad... credit_card    Computer             860xx    AZ         27.65 0           Jan-1985         735            739             1              NaN                    NaN                    3        0       13648     83.7%      9         f                   0.00      0.00          5863.1551866952    5833.84         5000.00         863.16        0.0                0.0        0.0                     Jan-2015     171.62          NaN          May-2019           714                  710                 0                          NaN                         1           Individual       NaN              NaN       NaN                       0              NaN          NaN         NaN         NaN         NaN         NaN         NaN                NaN          NaN     NaN         NaN         NaN        NaN      NaN              NaN    NaN         NaN          NaN                  NaN         NaN            NaN     0                        0           NaN                NaN                  NaN 

In [10]:
df_2007_2011.tail()

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
72176                                           NaN 2525 2525 225  36 months   9.33% 80.69  B   B3  NaN       < 1 year  RENT     110000 Not Verified Jun-2007 Does not meet the credit policy. Status:Fully Paid n   https://lendingclub.com/browse/loanDetail.actio... I need to pay $2,100 for fixing my Volvo :)  An... other Car repair bill 100xx NY  10   NaN NaN 710 714 NaN NaN NaN NaN NaN 0   NaN NaN f   0.00 0.00 2904.49882892     258.82 2525.00 379.50 0.0 0.0 0.0 Jun-2010 82.03  Jul-2010 May-2007 714 710 NaN NaN 1   Individual NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN                                                NaN                              
71623                                           NaN 6500 6500 0    36 months   8.38% 204.84 A   A5  NaN       < 1 year  NONE     NaN    Not Verified Jun-2007 Does not meet the credit policy. Status:Fully Paid n   https://lendingclub.com/browse/loanDetail.actio... Hi,   I'm buying  a used car. Anybody on facebo... other Buying a car    100xx NY  4    NaN NaN 740 744 NaN NaN NaN NaN NaN 0   NaN NaN f   0.00 0.00 7373.904961698404 0.00   6500.00 873.90 0.0 0.0 0.0 Jun-2010 205.32 Jul-2010 Aug-2007 724 720 NaN NaN 1   Individual NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN                                                NaN                              
70686                                           NaN 5000 5000 0    36 months   7.75% 156.11 A   A3  Homemaker 10+ years MORTGAGE 70000  Not Verified Jun-2007 Does not meet the credit policy. Status:Fully Paid n   https://lendingclub.com/browse/loanDetail.actio... I need to make several improvements around the ... other Aroundthehouse  068xx CT  8.81 NaN NaN 770 774 NaN NaN NaN NaN NaN 0   NaN NaN f   0.00 0.00 5619.762090469702 0.00   5000.00 619.76 0.0 0.0 0.0 Jun-2010 156.39 Jul-2010 Feb-2015 794 790 NaN NaN 1   Individual NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN                                                NaN                              
Total amount funded in policy code 1: 460296150 NaN NaN  NaN  NaN NaN        NaN     NaN    NaN NaN NaN       NaN       NaN      NaN    NaN          NaN      NaN                                                NaN NaN                                                NaN                                                NaN   NaN             NaN   NaN NaN  NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN  NaN  NaN               NaN    NaN     NaN    NaN NaN NaN NaN      NaN    NaN      NaN      NaN NaN NaN NaN NaN NaN        NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN                                                NaN                              

**NOTE: header = first row, footer = last two rows**

In [11]:
%%time
df_2007_2011 = pd.read_csv(dpath+'{}'.format(files[1]), header = 1, skipfooter = 2, engine = 'python')

CPU times: user 4.59 s, sys: 269 ms, total: 4.86 s
Wall time: 4.9 s


In [12]:
df_2007_2011.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501        NaN     5000.0       5000.0           4975.0   36 months   
1  1077430        NaN     2500.0       2500.0           2500.0   60 months   
2  1077175        NaN     2400.0       2400.0           2400.0   36 months   
3  1076863        NaN    10000.0      10000.0          10000.0   36 months   
4  1075358        NaN     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off   
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-2011   Fully Paid   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65          0.0   
1                  bike    309xx         GA   1.00          0.0   
2  real estate business    606xx         IL   8.72          0.0   
3              personel    917xx         CA  20.00          0.0   
4              Personal    972xx         OR  17.94          0.0   

  earliest_cr_line  fico_range_low  fico_range_high  inq_last_6mths  \
0         Jan-1985           735.0            739.0             1.0   
1         Apr-1999           740.0            744.0             5.0   
2         Nov-2001           735.0            739.0             2.0   
3         Feb-1996           690.0            694.0             1.0   
4         Jan-1996           695.0            699.0             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       3.0      0.0   
1                     NaN                     NaN       3.0      0.0   
2                     NaN                     NaN       2.0      0.0   
3                    35.0                     NaN      10.0      0.0   
4                    38.0                     NaN      15.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0    13648.0      83.7%        9.0                   f        0.0   
1     1687.0       9.4%        4.0                   f        0.0   
2     2956.0      98.5%       10.0                   f        0.0   
3     5598.0        21%       37.0                   f        0.0   
4    27783.0      53.9%       38.0                   f        0.0   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0            0.0   5863.1

In [13]:
df_2007_2011.tail()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
42531  73582        NaN     3500.0       3500.0            225.0   36 months   
42532  72998        NaN     1000.0       1000.0              0.0   36 months   
42533  72176        NaN     2525.0       2525.0            225.0   36 months   
42534  71623        NaN     6500.0       6500.0              0.0   36 months   
42535  70686        NaN     5000.0       5000.0              0.0   36 months   

      int_rate  installment grade sub_grade                   emp_title  \
42531   10.28%       113.39     C        C1                         NaN   
42532    9.64%        32.11     B        B4  Halping hands company inc.   
42533    9.33%        80.69     B        B3                         NaN   
42534    8.38%       204.84     A        A5                         NaN   
42535    7.75%       156.11     A        A3                   Homemaker   

      emp_length home_ownership  annual_inc verification_status   issue_d  \
42531   < 1 year           RENT    180000.0        Not Verified  Jun-2007   
42532   < 1 year           RENT     12000.0        Not Verified  Jun-2007   
42533   < 1 year           RENT    110000.0        Not Verified  Jun-2007   
42534   < 1 year           NONE         NaN        Not Verified  Jun-2007   
42535  10+ years       MORTGAGE     70000.0        Not Verified  Jun-2007   

                                             loan_status pymnt_plan  \
42531  Does not meet the credit policy. Status:Fully ...          n   
42532  Does not meet the credit policy. Status:Fully ...          n   
42533  Does not meet the credit policy. Status:Fully ...          n   
42534  Does not meet the credit policy. Status:Fully ...          n   
42535  Does not meet the credit policy. Status:Fully ...          n   

                                                     url  \
42531  https://lendingclub.com/browse/loanDetail.acti...   
42532  https://lendingclub.com/browse/loanDetail.acti...   
42533  https://lendingclub.com/browse/loanDetail.acti...   
42534  https://lendingclub.com/browse/loanDetail.acti...   
42535  https://lendingclub.com/browse/loanDetail.acti...   

                                                    desc purpose  \
42531  I am getting married on July 28 and will need ...   other   
42532  I would like to buy some new furniture in my a...   other   
42533  I need to pay $2,100 for fixing my Volvo :)  A...   other   
42534  Hi,   I'm buying  a used car. Anybody on faceb...   other   
42535  I need to make several improvements around the...   other   

                   title zip_code addr_state    dti  delinq_2yrs  \
42531  Wedding coming up    100xx         NY  10.00          NaN   
42532            delight    021xx         MA  10.00          NaN   
42533    Car repair bill    100xx         NY  10.00          NaN   
42534       Buying a car    100xx         NY   4.00          NaN   
42535     Aroundthehouse    068xx         CT   8.81          NaN   

      earliest_cr_line  fico_range_low  fico_range_high  inq_last_6mths  \
42531              NaN           685.0            689.0             NaN   
42532              NaN           695.0            699.0             NaN   
42533              NaN           710.0            714.0             NaN   
42534              NaN           740.0            744.0             NaN   
42535              NaN           770.0            774.0             NaN   

       mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
42531                     NaN                     NaN       NaN      NaN   
42532                     NaN                     NaN       NaN      NaN   
42533                     NaN                     NaN       NaN      NaN   
42534                     NaN                     NaN       NaN      NaN   
42535                     NaN                     NaN       NaN      NaN   

       revol_bal revol_util  total_acc initial_list_status  out_prncp  \
42531        0.0        NaN 

In [12]:
df_2007_2011.shape

(42536, 150)

In [14]:
cl_d2007 = df_2007_2011.columns.tolist()

In [15]:
cl_d2007[:6]

['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term']

[**back to top**](#toc)
<a id = '2012'></a>
#### 1.B. Check data file 2012-2013

In [16]:
%%time
df_2012_2013 = pd.read_csv(dpath+'{}'.format(files[2]), low_memory = False)

CPU times: user 14.7 s, sys: 1.2 s, total: 15.9 s
Wall time: 16 s


In [17]:
df_2012_2013.head()

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
id       member_id loan_amnt funded_amnt funded_amnt_inv term       int_rate installment grade sub_grade emp_title                           emp_length home_ownership annual_inc verification_status issue_d  loan_status pymnt_plan url                                                desc                                               purpose            title                                  zip_code addr_state dti   delinq_2yrs earliest_cr_line fico_range_low fico_range_high inq_last_6mths mths_since_last_delinq mths_since_last_record open_acc pub_rec revol_bal revol_util total_acc initial_list_status out_prncp out_prncp_inv total_pymnt      total_pymnt_inv total_rec_prncp total_rec_int total_rec_late_fee recoveries collection_recovery_fee last_pymnt_d last_pymnt_amnt next_pymnt_d last_credit_pull_d last_fico_range_high last_fico_range_low collections_12_mths_ex_med mths_since_last_major_derog policy_code application_type annual_inc_joint dti_joint verification_status_joint acc_now_delinq tot_coll_amt tot_cur_bal open_acc_6m open_act_il open_il_12m open_il_24m mths_since_rcnt_il total_bal_il il_util open_rv_12m open_rv_24m max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m acc_open_past_24mths avg_cur_bal bc_open_to_buy bc_util chargeoff_within_12_mths delinq_amnt mo_sin_old_il_acct mo_sin_old_rev_tl_op mo_sin_rcnt_rev_tl_op mo_sin_rcnt_tl mort_acc mths_since_recent_bc mths_since_recent_bc_dlq mths_since_recent_inq mths_since_recent_revol_delinq num_accts_ever_120_pd num_actv_bc_tl num_actv_rev_tl num_bc_sats num_bc_tl num_il_tl num_op_rev_tl num_rev_accts num_rev_tl_bal_gt_0 num_sats num_tl_120dpd_2m num_tl_30dpd num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit total_il_high_credit_limit revol_bal_joint sec_app_fico_range_low sec_app_fico_range_high sec_app_earliest_cr_line sec_app_inq_last_6mths sec_app_mort_acc sec_app_open_acc sec_app_revol_util sec_app_open_act_il sec_app_num_rev_accts sec_app_chargeoff_within_12_mths sec_app_collections_12_mths_ex_med sec_app_mths_since_last_major_derog hardship_flag hardship_type hardship_reason hardship_status deferral_term hardship_amount hardship_start_date hardship_end_date payment_plan_start_date hardship_length hardship_dpd hardship_loan_status orig_projected_additional_accrued_interest hardship_payoff_balance_amount hardship_last_payment_amount debt_settlement_flag debt_settlement_flag_date settlement_status settlement_date settlement_amount settlement_percentage                                    settlement_term                              
10129454 NaN       12000     12000       12000            36 months  10.99%  392.81      B     B2        Project Manager                     4 years    RENT           60000      Not Verified        Dec-2013 Fully Paid  n          https://lendingclub.com/browse/loanDetail.actio...   Borrower added on 12/31/13 > I would like to ... debt_consolidation No Regrets                             281xx    NC         4.62  0           Dec-2009         720            724             1              NaN                    NaN                    15       0       7137      24%        18        f                   0.00      0.00          13988.6099956242 13988.61        12000.00        1988.61       0.0                0.0        0.0                     Apr-2016     3775.55         NaN          Aug-2018           569                  565                 0                          NaN                         1           Individual       NaN              NaN       NaN                       0              0            7137        NaN         NaN         NaN         NaN         NaN                NaN          NaN     NaN         NaN         NaN        NaN      29700            NaN    NaN         NaN          8                    476         15216          15.9    0      

In [18]:
df_2012_2013.tail()

Notes offered by Prospectus (https://www.lendingclub.com/info/prospectus.action)
1059224                                          NaN 35000 35000 35000  36 months  15.96% 1229.81 C   C5  Tom and Holly Gores 3 years MORTGAGE 160000 Source Verified Jan-2012 Fully Paid  n   https://lendingclub.com/browse/loanDetail.actio... NaN                                                small_business Small Business Loan 922xx CA  4.9   0   Sep-2000 720 724 1   NaN NaN 9   0   23665 62.4% 17  f   0.00 0.00 44272.9399842368  44272.94 35000.00 9272.94 0.0 0.0    0.0  Dec-2014 1244.72 NaN Dec-2014 739 735 0   NaN 1   Individual NaN NaN NaN 0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN 0   0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN 0   0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN                                                NaN                              
1058722                                          NaN 12000 12000 12000  36 months  16.29% 423.61  D   D1  NaN                 NaN     MORTGAGE 35000  Source Verified Jan-2012 Charged Off n   https://lendingclub.com/browse/loanDetail.actio...   Borrower added on 12/06/11 > need to pay off ... other          others              770xx TX  12.93 0   Aug-2001 675 679 0   NaN NaN 14  0   15006 93.2% 27  f   0.00 0.00 2077.189999999999 2077.19  1063.61  629.79  0.0 383.79 3.99 Apr-2012 423.61  NaN Oct-2016 584 580 0   NaN 1   Individual NaN NaN NaN 0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN 0   0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN 0   0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN                                                NaN                              
1058291                                          NaN 12000 7775  7775   60 months  15.27% 186.08  C   C4  NaN                 7 years RENT     50000  Verified        Jan-2012 Fully Paid  n   https://lendingclub.com/browse/loanDetail.actio...   Borrower added on 12/06/11 > Want to close do... credit_card    refinance           220xx VA  5.5   0   Feb-2003 715 719 1   NaN NaN 13  0   7008  37.9% 25  f   0.00 0.00 11163.9482553723  11163.95 7775.00  3388.95 0.0 0.0    0.0  Dec-2016 185.23  NaN Jan-2019 569 565 0   NaN 1   Individual NaN NaN NaN 0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN 0   0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN 0   0   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN N   NaN NaN NaN NaN NaN                                                NaN                              
Total amount funded in policy code 1: 2700702175 NaN NaN   NaN   NaN   NaN        NaN     NaN     NaN NaN NaN                 NaN     NaN      NaN    NaN             NaN      NaN         NaN NaN                                                NaN                                                NaN            NaN                 NaN   NaN NaN   NaN NaN      NaN NaN NaN NaN NaN NaN NaN NaN   NaN   NaN NaN NaN  NaN  NaN               NaN      NaN      NaN     NaN NaN    NaN  NaN      NaN     NaN NaN      NaN NaN NaN NaN NaN NaN        NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN                                                NaN          

**NOTE: file format consistent with first data file: header = first row, footer = last two rows**

In [19]:
%%time
df_2012_2013 = pd.read_csv(dpath+'{}'.format(files[2]), header = 1, skipfooter = 2, engine = 'python' )

CPU times: user 24.1 s, sys: 1.27 s, total: 25.4 s
Wall time: 25.8 s


In [20]:
df_2012_2013.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  10129454        NaN      12000        12000          12000.0   36 months   
1  10149488        NaN       4800         4800           4800.0   36 months   
2  10148122        NaN      12000        12000          12000.0   36 months   
3  10149342        NaN      27050        27050          27050.0   36 months   
4  10129477        NaN      14000        14000          14000.0   36 months   

  int_rate  installment grade sub_grade                             emp_title  \
0   10.99%       392.81     B        B2                       Project Manager   
1   10.99%       157.13     B        B2                   Surgical Technician   
2    7.62%       373.94     A        A3                      Systems Engineer   
3   10.99%       885.46     B        B2   Team Leadern Customer Ops & Systems   
4   12.85%       470.71     B        B4  Assistant Director - Human Resources   

  emp_length home_ownership  annual_inc verification_status   issue_d  \
0    4 years           RENT     60000.0        Not Verified  Dec-2013   
1    2 years       MORTGAGE     39600.0     Source Verified  Dec-2013   
2    3 years       MORTGAGE     96500.0        Not Verified  Dec-2013   
3  10+ years            OWN     55000.0            Verified  Dec-2013   
4    4 years           RENT     88000.0        Not Verified  Dec-2013   

  loan_status pymnt_plan                                                url  \
0  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   
1  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   
2  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   
3  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   
4  Fully Paid          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc             purpose  \
0    Borrower added on 12/31/13 > I would like to...  debt_consolidation   
1    Borrower added on 12/31/13 > Just bought a h...    home_improvement   
2    Borrower added on 12/31/13 > Bought a new ho...  debt_consolidation   
3    Borrower added on 12/31/13 > Combining high ...  debt_consolidation   
4                                                NaN  debt_consolidation   

                                    title zip_code addr_state    dti  \
0                              No Regrets    281xx         NC   4.62   
1                           For The House    782xx         TX   2.49   
2  Debt Consolidation and Credit Transfer    782xx         TX  12.61   
3                      Debt Consolidation    481xx         MI  22.87   
4                      Debt consolidation    282xx         NC  10.02   

   delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
0            0         Dec-2009             720              724   
1            0         Aug-1995             755              759   
2            0         Sep-2003             705              709   
3            0         Oct-1986             730              734   
4            1         Jun-1988             670              674   

   inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  \
0               1                     NaN                     NaN        15   
1               2                     NaN                     NaN         3   
2               0                     NaN                     NaN        17   
3               0                     NaN                     NaN        14   
4               0                    16.0                   115.0         6   

   pub_rec  revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0        0       7137        24%         18                   f        0.0   
1        0       4136      16.1%          8                   w        0.0   
2        0      13248      55.7%         30                   f        0.0   
3        0      36638      61.2%         27    

In [21]:
df_2012_2013.tail()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
188176  1062400        NaN      20500        20500          20500.0   
188177  1059394        NaN      15000        15000          15000.0   
188178  1059224        NaN      35000        35000          35000.0   
188179  1058722        NaN      12000        12000          12000.0   
188180  1058291        NaN      12000         7775           7775.0   

              term int_rate  installment grade sub_grade            emp_title  \
188176   36 months   16.77%       728.54     D        D2                  NaN   
188177   36 months   15.27%       521.97     C        C4                  NaN   
188178   36 months   15.96%      1229.81     C        C5  Tom and Holly Gores   
188179   36 months   16.29%       423.61     D        D1                  NaN   
188180   60 months   15.27%       186.08     C        C4                  NaN   

       emp_length home_ownership  annual_inc verification_status   issue_d  \
188176    7 years           RENT     60000.0     Source Verified  Jan-2012   
188177    3 years           RENT     57600.0     Source Verified  Jan-2012   
188178    3 years       MORTGAGE    160000.0     Source Verified  Jan-2012   
188179        NaN       MORTGAGE     35000.0     Source Verified  Jan-2012   
188180    7 years           RENT     50000.0            Verified  Jan-2012   

        loan_status pymnt_plan  \
188176   Fully Paid          n   
188177   Fully Paid          n   
188178   Fully Paid          n   
188179  Charged Off          n   
188180   Fully Paid          n   

                                                      url  \
188176  https://lendingclub.com/browse/loanDetail.acti...   
188177  https://lendingclub.com/browse/loanDetail.acti...   
188178  https://lendingclub.com/browse/loanDetail.acti...   
188179  https://lendingclub.com/browse/loanDetail.acti...   
188180  https://lendingclub.com/browse/loanDetail.acti...   

                                                     desc             purpose  \
188176    Borrower added on 12/08/11 > I will be payin...  debt_consolidation   
188177                                                NaN      small_business   
188178                                                NaN      small_business   
188179    Borrower added on 12/06/11 > need to pay off...               other   
188180    Borrower added on 12/06/11 > Want to close d...         credit_card   

                      title zip_code addr_state    dti  delinq_2yrs  \
188176          Payoff Loan    100xx         NY  16.40            1   
188177        business loan    900xx         CA   8.35            2   
188178  Small Business Loan    922xx         CA   4.90            0   
188179               others    770xx         TX  12.93            0   
188180            refinance    220xx         VA   5.50            0   

       earliest_cr_line  fico_range_low  fico_range_high  inq_last_6mths  \
188176         Oct-1988             700              704               3   
188177         Jan-2004             680              684               1   
188178         Sep-2000             720              724               1   
188179         Aug-2001             675              679               0   
188180         Feb-2003             715              719               1   

        mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
188176                    20.0                     NaN        10        0   
188177                    10.0                     NaN        18        0   
188178                     NaN                     NaN         9        0   
188179                     NaN                     NaN        14        0   
188180                     NaN                     NaN        13        0   

        revol_bal revol_util  total_acc initial_list_status  out_prncp  \
188176      15417      58.4%         20                   f        0.0   
188177       8897      33.1%         30                   f        0.0   
18

In [22]:
df_2012_2013.shape

(188181, 150)

In [23]:
cl_d2012 = df_2012_2013.columns.tolist()

In [24]:
cl_d2007 == cl_d2012

True

**NOTE: good sign: columns from both the first and second data files are identical**

[**back to top**](#toc)
<a id = '2016q1'></a>
#### 1.C. Check data file 2016Q1

In [29]:
%%time
df_2016q1 = pd.read_csv(dpath+'{}'.format(files[5]), header = 1, skipfooter = 2, engine = 'python' )

CPU times: user 17.1 s, sys: 378 ms, total: 17.4 s
Wall time: 17.5 s


In [30]:
df_2016q1.shape

(133887, 150)

In [31]:
df_2016q1.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  74121690        NaN       6000         6000           6000.0   36 months   
1  74724861        NaN      21000        21000          21000.0   60 months   
2  74826201        NaN       7200         7200           7200.0   36 months   
3  75061311        NaN      12000        12000          12000.0   60 months   
4  75091735        NaN      11425        11425          11425.0   36 months   

  int_rate  installment grade sub_grade        emp_title emp_length  \
0   12.99%       202.14     C        C2         Salesman    9 years   
1   19.53%       550.90     D        D5  Human Resources  10+ years   
2    5.32%       216.83     A        A1         Mechanic  10+ years   
3   11.99%       266.88     C        C1         Mechanic  10+ years   
4   19.53%       421.87     D        D5            Nurse    5 years   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0       MORTGAGE     43000.0        Not Verified  Mar-2016   Fully Paid   
1       MORTGAGE     65000.0            Verified  Mar-2016      Current   
2       MORTGAGE     49000.0     Source Verified  Mar-2016   Fully Paid   
3       MORTGAGE     49000.0        Not Verified  Mar-2016  Charged Off   
4           RENT     26000.0     Source Verified  Mar-2016   Fully Paid   

  pymnt_plan                                                url desc  \
0          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
1          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
2          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
3          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   
4          n  https://lendingclub.com/browse/loanDetail.acti...  NaN   

              purpose                    title zip_code addr_state    dti  \
0  debt_consolidation       Debt consolidation    211xx         MD  14.40   
1         credit_card  Credit card refinancing    430xx         OH  43.83   
2  debt_consolidation       Debt consolidation    554xx         MN  19.05   
3         credit_card                      NaN    333xx         FL   6.59   
4      major_purchase           Major purchase    328xx         FL  35.56   

   delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
0            0         May-2006             675              679   
1            1         Mar-2000             710              714   
2            0         Dec-2001             750              754   
3            1         Dec-1999             670              674   
4            0         Sep-2008             730              734   

   inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  \
0             1.0                     NaN                     NaN        16   
1             2.0                    15.0                     NaN        13   
2             1.0                    41.0                     NaN        11   
3             1.0                    19.0                     NaN        10   
4             0.0                     NaN                     NaN        11   

   pub_rec  revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0        0      15055        63%         21                   f       0.00   
1        0     124946      65.5%         25                   w    8123.69   
2        0       9309      18.4%         36                   w       0.00   
3        0      12152      50.6%         29                   w       0.00   
4        0       2096       7.2%         18                   f       0.00   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0           0.00   7268.153165          7268.15          6000.00   
1        8123.69  23643.130000         23643.13         12876.31   
2           0.00   7223.410000          7223.41          7200.00   
3           0.00   8448.900000          8448.90          5038.27   
4           0.00  14331.034913         14331.03         11425.00   

In [32]:
df_2016q1.tail()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
133882  66055600        NaN       6000         6000           6000.0   
133883  65854936        NaN       6000         6000           6000.0   
133884  66141895        NaN      14400        14400          14400.0   
133885  65673209        NaN      34050        34050          34050.0   
133886  65744272        NaN       5000         5000           5000.0   

              term int_rate  installment grade sub_grade        emp_title  \
133882   36 months    9.17%       191.28     B        B2              NaN   
133883   36 months    7.89%       187.72     A        A5  Warehouse Clerk   
133884   60 months   13.18%       328.98     C        C3       Meatcutter   
133885   36 months   15.41%      1187.21     D        D1       Supervisor   
133886   36 months   11.22%       164.22     B        B5        carpenter   

       emp_length home_ownership  annual_inc verification_status   issue_d  \
133882        NaN           RENT     32640.0        Not Verified  Jan-2016   
133883   < 1 year            OWN     38000.0     Source Verified  Jan-2016   
133884  10+ years           RENT     47000.0            Verified  Jan-2016   
133885  10+ years       MORTGAGE     87800.0     Source Verified  Jan-2016   
133886    7 years       MORTGAGE     65000.0     Source Verified  Jan-2016   

        loan_status pymnt_plan  \
133882   Fully Paid          n   
133883   Fully Paid          n   
133884  Charged Off          n   
133885   Fully Paid          n   
133886   Fully Paid          n   

                                                      url desc  \
133882  https://lendingclub.com/browse/loanDetail.acti...  NaN   
133883  https://lendingclub.com/browse/loanDetail.acti...  NaN   
133884  https://lendingclub.com/browse/loanDetail.acti...  NaN   
133885  https://lendingclub.com/browse/loanDetail.acti...  NaN   
133886  https://lendingclub.com/browse/loanDetail.acti...  NaN   

                   purpose                    title zip_code addr_state  \
133882  debt_consolidation       Debt consolidation    600xx         IL   
133883         credit_card  Credit card refinancing    432xx         OH   
133884         credit_card  Credit card refinancing    531xx         WI   
133885         credit_card  Credit card refinancing    212xx         MD   
133886    home_improvement         Home improvement    201xx         VA   

          dti  delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
133882  22.76            0         Sep-1997             695              699   
133883  12.35            0         Feb-2006             695              699   
133884  19.64            0         Oct-1976             670              674   
133885  12.10            1         Nov-2005             680              684   
133886   3.10            0         Jul-2005             665              669   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
133882             1.0                     NaN                   101.0   
133883             0.0                     NaN                     NaN   
133884             4.0                    48.0                     NaN   
133885             1.0                     8.0                     NaN   
133886             1.0                     NaN                    77.0   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
133882         9        1       6898      37.1%         17   
133883        10        0       7867      36.8%         14   
133884         7        0      10164        56%         19   
133885        14        0      25473      53.2%         21   
133886         6        1       5763        48%          8   

       initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
133882                   w        0.0            0.0   6903.429166   
133883                   w        0.0            0.0   6755.041951   
133884                   w        0.0            0.0  10368.610000   
133885           

In [33]:
def compare_df_columns(df1, df2):
    cl_df1 = df1.columns.tolist()
    cl_df2 = df2.columns.tolist()
    
    print ('Number of columns: first df: {}, second df: {}'.format(len(cl_df1), len(cl_df2)))
    print ('Columns from both dfs are identical:', cl_df1 == cl_df2)

In [34]:
compare_df_columns(df_2016q1, df_2007_2011)

Number of columns: first df: 150, second df: 150
Columns from both dfs are identical: True


[**back to top**](#toc)

<a id = '2019q3'></a>
#### 1.D. Check data file 2019Q3

In [35]:
%%time
df_2019q3 = pd.read_csv(dpath+'{}'.format(files[-1]), header = 1, skipfooter = 2, engine = 'python' )

CPU times: user 18.7 s, sys: 416 ms, total: 19.1 s
Wall time: 19.1 s


In [36]:
df_2019q3.shape

(143035, 150)

In [37]:
df_2019q3.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  158303403        NaN      12000        12000            12000   36 months   
1  158628181        NaN      20000        20000            20000   36 months   
2  159231951        NaN      18000        18000            18000   60 months   
3  159289397        NaN      10000        10000            10000   36 months   
4  159296485        NaN       7000         7000             7000   36 months   

  int_rate  installment grade sub_grade              emp_title emp_length  \
0    8.19%       377.09     A        A4                Teacher  10+ years   
1    8.81%       634.23     A        A5         Remodel expert    3 years   
2   13.08%       410.30     B        B5  Operations Supervisor    3 years   
3   10.33%       324.23     B        B1                    NaN   < 1 year   
4   18.62%       255.25     D        D1         Street sweeper    6 years   

  home_ownership  annual_inc verification_status   issue_d loan_status  \
0           RENT     84000.0     Source Verified  Sep-2019     Current   
1           RENT     44000.0        Not Verified  Sep-2019     Current   
2           RENT     50000.0     Source Verified  Sep-2019     Current   
3           RENT     59300.0     Source Verified  Sep-2019     Current   
4           RENT     55000.0            Verified  Sep-2019     Current   

  pymnt_plan                                                url  desc  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   NaN   
1          n  https://lendingclub.com/browse/loanDetail.acti...   NaN   
2          n  https://lendingclub.com/browse/loanDetail.acti...   NaN   
3          n  https://lendingclub.com/browse/loanDetail.acti...   NaN   
4          n  https://lendingclub.com/browse/loanDetail.acti...   NaN   

              purpose               title zip_code addr_state    dti  \
0  debt_consolidation  Debt consolidation    937xx         CA  10.86   
1  debt_consolidation  Debt consolidation    891xx         NV  13.72   
2  debt_consolidation  Debt consolidation    750xx         TX  14.47   
3  debt_consolidation  Debt consolidation    600xx         IL  19.81   
4  debt_consolidation  Debt consolidation    902xx         CA  16.98   

   delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
0            1         Jan-1996             680              684   
1            0         Feb-2006             745              749   
2            0         Sep-2010             660              664   
3            0         Jan-2005             720              724   
4            2         Sep-2010             670              674   

   inq_last_6mths  mths_since_last_delinq  mths_since_last_record  open_acc  \
0               0                    17.0                     NaN         5   
1               0                     NaN                     NaN         6   
2               0                     NaN                     NaN        13   
3               0                    41.0                    90.0        17   
4               0                    17.0                     NaN         8   

   pub_rec  revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0        0      22489      71.6%         15                   w   11713.00   
1        0       5435      21.8%         12                   w   18947.67   
2        0      13973      85.7%         13                   w   17785.90   
3        1       3121       8.9%         20                   w    9761.85   
4        0       4546      58.3%         19                   w    6853.37   

   out_prncp_inv  total_pymnt  total_pymnt_inv  total_rec_prncp  \
0       11713.00       371.63           371.63           287.00   
1       18947.67      1258.67          1258.67          1052.33   
2       17785.90       397.22           397.22           214.10   
3        9761.85       318.49           318.49           238.15   
4        6853.37       248.01           248.01           146.

In [38]:
df_2019q3.tail()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
143030  153709478        NaN      23300        23300            23300   
143031  153714644        NaN      20000        20000            20000   
143032  152956054        NaN      26575        26575            26575   
143033  153350995        NaN      10000        10000            10000   
143034  153062400        NaN      20000        20000            20000   

              term int_rate  installment grade sub_grade  \
143030   60 months   22.50%       650.17     D        D3   
143031   60 months   20.00%       529.88     D        D2   
143032   60 months   15.24%       635.58     C        C2   
143033   36 months    8.81%       317.12     A        A5   
143034   60 months   22.50%       558.08     D        D3   

                     emp_title emp_length home_ownership  annual_inc  \
143030  Service Representative  10+ years       MORTGAGE     60000.0   
143031           Teen director    7 years       MORTGAGE     55000.0   
143032                     NaN   < 1 year           RENT    100000.0   
143033    Program Coordinator   10+ years       MORTGAGE     52116.0   
143034                   Clerk  10+ years           RENT     48000.0   

       verification_status   issue_d loan_status pymnt_plan  \
143030        Not Verified  Jul-2019     Current          n   
143031        Not Verified  Jul-2019     Current          n   
143032        Not Verified  Jul-2019     Current          n   
143033            Verified  Jul-2019     Current          n   
143034            Verified  Jul-2019     Current          n   

                                                      url  desc  \
143030  https://lendingclub.com/browse/loanDetail.acti...   NaN   
143031  https://lendingclub.com/browse/loanDetail.acti...   NaN   
143032  https://lendingclub.com/browse/loanDetail.acti...   NaN   
143033  https://lendingclub.com/browse/loanDetail.acti...   NaN   
143034  https://lendingclub.com/browse/loanDetail.acti...   NaN   

                   purpose               title zip_code addr_state    dti  \
143030  debt_consolidation  Debt consolidation    336xx         FL  30.00   
143031    home_improvement    Home improvement    681xx         NE  25.73   
143032  debt_consolidation  Debt consolidation    900xx         CA  14.15   
143033  debt_consolidation  Debt consolidation    921xx         CA  22.45   
143034  debt_consolidation  Debt consolidation    463xx         IN  27.83   

        delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
143030            0         Aug-2000             695              699   
143031            0         Jul-1992             730              734   
143032            0         Jul-2002             715              719   
143033            0         Feb-1998             685              689   
143034            2         Jun-2009             665              669   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
143030               0                     NaN                     NaN   
143031               0                     NaN                     NaN   
143032               0                    35.0                     NaN   
143033               0                    48.0                     NaN   
143034               1                    14.0                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
143030        19        0      24776      37.9%         31   
143031         7        0       9773      82.1%         13   
143032        10        0      34813      49.6%         21   
143033        14        0       6847      21.1%         30   
143034         8        0       4960      42.4%         21   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
143030                   w   22422.52       22422.52      2571.56   
143031                   w   19193.94       19193.94      2097.30   
143032                   w   25359.78       25359.78      2519.82   
1430

In [39]:
compare_df_columns(df_2007_2011, df_2019q3)

Number of columns: first df: 150, second df: 150
Columns from both dfs are identical: True


**NOTE: Data files appear to be consistent in format and column layout.**

[**back to top**](#toc)

<a id = 'agg'></a>
### 2. Aggregate data from all files into one

In [42]:
%%time
df_list = []
for i in range(1, len(files)):
    df = pd.read_csv((dpath+'{}'.format(files[i])), header = 1, skipfooter = 2, engine = 'python')
    df_list.append(df)

CPU times: user 6min 22s, sys: 12.4 s, total: 6min 35s
Wall time: 6min 44s


In [45]:
len(df_list)

19

In [46]:
%%time
df = pd.concat(df_list, ignore_index=True)

CPU times: user 14.4 s, sys: 7.34 s, total: 21.7 s
Wall time: 19.5 s


In [47]:
df.shape

(2650518, 150)

In [48]:
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501        NaN     5000.0       5000.0           4975.0   36 months   
1  1077430        NaN     2500.0       2500.0           2500.0   60 months   
2  1077175        NaN     2400.0       2400.0           2400.0   36 months   
3  1076863        NaN    10000.0      10000.0          10000.0   36 months   
4  1075358        NaN     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off   
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-2011   Fully Paid   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65          0.0   
1                  bike    309xx         GA   1.00          0.0   
2  real estate business    606xx         IL   8.72          0.0   
3              personel    917xx         CA  20.00          0.0   
4              Personal    972xx         OR  17.94          0.0   

  earliest_cr_line  fico_range_low  fico_range_high  inq_last_6mths  \
0         Jan-1985           735.0            739.0             1.0   
1         Apr-1999           740.0            744.0             5.0   
2         Nov-2001           735.0            739.0             2.0   
3         Feb-1996           690.0            694.0             1.0   
4         Jan-1996           695.0            699.0             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       3.0      0.0   
1                     NaN                     NaN       3.0      0.0   
2                     NaN                     NaN       2.0      0.0   
3                    35.0                     NaN      10.0      0.0   
4                    38.0                     NaN      15.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0    13648.0      83.7%        9.0                   f        0.0   
1     1687.0       9.4%        4.0                   f        0.0   
2     2956.0      98.5%       10.0                   f        0.0   
3     5598.0        21%       37.0                   f        0.0   
4    27783.0      53.9%       38.0                   f        0.0   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0            0.0   5863.1

In [49]:
df_2007_2011.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501        NaN     5000.0       5000.0           4975.0   36 months   
1  1077430        NaN     2500.0       2500.0           2500.0   60 months   
2  1077175        NaN     2400.0       2400.0           2400.0   36 months   
3  1076863        NaN    10000.0      10000.0          10000.0   36 months   
4  1075358        NaN     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off   
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-2011   Fully Paid   

  pymnt_plan                                                url  \
0          n  https://lendingclub.com/browse/loanDetail.acti...   
1          n  https://lendingclub.com/browse/loanDetail.acti...   
2          n  https://lendingclub.com/browse/loanDetail.acti...   
3          n  https://lendingclub.com/browse/loanDetail.acti...   
4          n  https://lendingclub.com/browse/loanDetail.acti...   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state    dti  delinq_2yrs  \
0              Computer    860xx         AZ  27.65          0.0   
1                  bike    309xx         GA   1.00          0.0   
2  real estate business    606xx         IL   8.72          0.0   
3              personel    917xx         CA  20.00          0.0   
4              Personal    972xx         OR  17.94          0.0   

  earliest_cr_line  fico_range_low  fico_range_high  inq_last_6mths  \
0         Jan-1985           735.0            739.0             1.0   
1         Apr-1999           740.0            744.0             5.0   
2         Nov-2001           735.0            739.0             2.0   
3         Feb-1996           690.0            694.0             1.0   
4         Jan-1996           695.0            699.0             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       3.0      0.0   
1                     NaN                     NaN       3.0      0.0   
2                     NaN                     NaN       2.0      0.0   
3                    35.0                     NaN      10.0      0.0   
4                    38.0                     NaN      15.0      0.0   

   revol_bal revol_util  total_acc initial_list_status  out_prncp  \
0    13648.0      83.7%        9.0                   f        0.0   
1     1687.0       9.4%        4.0                   f        0.0   
2     2956.0      98.5%       10.0                   f        0.0   
3     5598.0        21%       37.0                   f        0.0   
4    27783.0      53.9%       38.0                   f        0.0   

   out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
0            0.0   5863.1

In [50]:
df.tail()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
2650513  153709478        NaN    23300.0      23300.0          23300.0   
2650514  153714644        NaN    20000.0      20000.0          20000.0   
2650515  152956054        NaN    26575.0      26575.0          26575.0   
2650516  153350995        NaN    10000.0      10000.0          10000.0   
2650517  153062400        NaN    20000.0      20000.0          20000.0   

               term int_rate  installment grade sub_grade  \
2650513   60 months   22.50%       650.17     D        D3   
2650514   60 months   20.00%       529.88     D        D2   
2650515   60 months   15.24%       635.58     C        C2   
2650516   36 months    8.81%       317.12     A        A5   
2650517   60 months   22.50%       558.08     D        D3   

                      emp_title emp_length home_ownership  annual_inc  \
2650513  Service Representative  10+ years       MORTGAGE     60000.0   
2650514           Teen director    7 years       MORTGAGE     55000.0   
2650515                     NaN   < 1 year           RENT    100000.0   
2650516    Program Coordinator   10+ years       MORTGAGE     52116.0   
2650517                   Clerk  10+ years           RENT     48000.0   

        verification_status   issue_d loan_status pymnt_plan  \
2650513        Not Verified  Jul-2019     Current          n   
2650514        Not Verified  Jul-2019     Current          n   
2650515        Not Verified  Jul-2019     Current          n   
2650516            Verified  Jul-2019     Current          n   
2650517            Verified  Jul-2019     Current          n   

                                                       url desc  \
2650513  https://lendingclub.com/browse/loanDetail.acti...  NaN   
2650514  https://lendingclub.com/browse/loanDetail.acti...  NaN   
2650515  https://lendingclub.com/browse/loanDetail.acti...  NaN   
2650516  https://lendingclub.com/browse/loanDetail.acti...  NaN   
2650517  https://lendingclub.com/browse/loanDetail.acti...  NaN   

                    purpose               title zip_code addr_state    dti  \
2650513  debt_consolidation  Debt consolidation    336xx         FL  30.00   
2650514    home_improvement    Home improvement    681xx         NE  25.73   
2650515  debt_consolidation  Debt consolidation    900xx         CA  14.15   
2650516  debt_consolidation  Debt consolidation    921xx         CA  22.45   
2650517  debt_consolidation  Debt consolidation    463xx         IN  27.83   

         delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
2650513          0.0         Aug-2000           695.0            699.0   
2650514          0.0         Jul-1992           730.0            734.0   
2650515          0.0         Jul-2002           715.0            719.0   
2650516          0.0         Feb-1998           685.0            689.0   
2650517          2.0         Jun-2009           665.0            669.0   

         inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
2650513             0.0                     NaN                     NaN   
2650514             0.0                     NaN                     NaN   
2650515             0.0                    35.0                     NaN   
2650516             0.0                    48.0                     NaN   
2650517             1.0                    14.0                     NaN   

         open_acc  pub_rec  revol_bal revol_util  total_acc  \
2650513      19.0      0.0    24776.0      37.9%       31.0   
2650514       7.0      0.0     9773.0      82.1%       13.0   
2650515      10.0      0.0    34813.0      49.6%       21.0   
2650516      14.0      0.0     6847.0      21.1%       30.0   
2650517       8.0      0.0     4960.0      42.4%       21.0   

        initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
2650513                   w   22422.52       22422.52      2571.56   
2650514                   w   19193.94       19193.94      2097.30   
2650515                

In [51]:
df_2019q3.tail()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
143030  153709478        NaN      23300        23300            23300   
143031  153714644        NaN      20000        20000            20000   
143032  152956054        NaN      26575        26575            26575   
143033  153350995        NaN      10000        10000            10000   
143034  153062400        NaN      20000        20000            20000   

              term int_rate  installment grade sub_grade  \
143030   60 months   22.50%       650.17     D        D3   
143031   60 months   20.00%       529.88     D        D2   
143032   60 months   15.24%       635.58     C        C2   
143033   36 months    8.81%       317.12     A        A5   
143034   60 months   22.50%       558.08     D        D3   

                     emp_title emp_length home_ownership  annual_inc  \
143030  Service Representative  10+ years       MORTGAGE     60000.0   
143031           Teen director    7 years       MORTGAGE     55000.0   
143032                     NaN   < 1 year           RENT    100000.0   
143033    Program Coordinator   10+ years       MORTGAGE     52116.0   
143034                   Clerk  10+ years           RENT     48000.0   

       verification_status   issue_d loan_status pymnt_plan  \
143030        Not Verified  Jul-2019     Current          n   
143031        Not Verified  Jul-2019     Current          n   
143032        Not Verified  Jul-2019     Current          n   
143033            Verified  Jul-2019     Current          n   
143034            Verified  Jul-2019     Current          n   

                                                      url  desc  \
143030  https://lendingclub.com/browse/loanDetail.acti...   NaN   
143031  https://lendingclub.com/browse/loanDetail.acti...   NaN   
143032  https://lendingclub.com/browse/loanDetail.acti...   NaN   
143033  https://lendingclub.com/browse/loanDetail.acti...   NaN   
143034  https://lendingclub.com/browse/loanDetail.acti...   NaN   

                   purpose               title zip_code addr_state    dti  \
143030  debt_consolidation  Debt consolidation    336xx         FL  30.00   
143031    home_improvement    Home improvement    681xx         NE  25.73   
143032  debt_consolidation  Debt consolidation    900xx         CA  14.15   
143033  debt_consolidation  Debt consolidation    921xx         CA  22.45   
143034  debt_consolidation  Debt consolidation    463xx         IN  27.83   

        delinq_2yrs earliest_cr_line  fico_range_low  fico_range_high  \
143030            0         Aug-2000             695              699   
143031            0         Jul-1992             730              734   
143032            0         Jul-2002             715              719   
143033            0         Feb-1998             685              689   
143034            2         Jun-2009             665              669   

        inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
143030               0                     NaN                     NaN   
143031               0                     NaN                     NaN   
143032               0                    35.0                     NaN   
143033               0                    48.0                     NaN   
143034               1                    14.0                     NaN   

        open_acc  pub_rec  revol_bal revol_util  total_acc  \
143030        19        0      24776      37.9%         31   
143031         7        0       9773      82.1%         13   
143032        10        0      34813      49.6%         21   
143033        14        0       6847      21.1%         30   
143034         8        0       4960      42.4%         21   

       initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
143030                   w   22422.52       22422.52      2571.56   
143031                   w   19193.94       19193.94      2097.30   
143032                   w   25359.78       25359.78      2519.82   
1430

In [46]:
%%time
df.to_csv('./data/master_loan_datav2019q3.csv', index = False)

CPU times: user 7min 1s, sys: 7.6 s, total: 7min 9s
Wall time: 7min 13s


[**back to top**](#toc)

<a id = 'notes'></a>

## 3. Notes

1. Individual data files:
    - header = 1, skipfooter = 2, engine = 'python'
    - columns are aligned with each other
    
2. Raw data: > 2.6 million data points, 150 features
    - too big for local machine to train, consider cloud service, e.g., AWS
    - may start with sample datasets on local machine for ideas
---
**End of current notebook**

**[back to top](#toc)**   

---
**Previous notebook**: **[part 0a - webScrape-dataCollection](proj-classification-loanDefault-p0a-webScrape-dataCollection-max-v2019Dec.ipynb)**

**Next notebook**: **[part 1 - EDA](proj-classification-loanDefault-p1-EDA-max-v2020Jan.ipynb)**